### PR TITLE
Support `GROUP BY column_number` 

### PR DESCRIPTION
**Pull request by dt**
[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3597)

### CODE REVIEW

**Comment by petermattis** [comment]
LGTMf

---

Reviewed 3 of 3 files at r1, 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/group.go, line 53 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7DulQ4S48glBi98VSe:-K7DulQ51-NRi6vOuigQ:834179476) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/group.go#L53)):*
Can't you skip `resolveQNames` and normalization in this case, as those steps will already have been done. (i.e. `continue` here).

---

*[sql/scan.go, line 524 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7DtvAYuy9ViaXNcqaV:-K7DtvAYuy9ViaXNcqaW:280349512) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd9fe4ae0d2383fbeaa7e842bb30fdcbbf3449e/sql/scan.go#L524)):*
`s/colOffset/colIndex/g`. You're using the term `index` below. I think it makes more sense in the method name as well.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-:-K7Dv6gxALwd8xwkeks3:-492975155)*

**Comment by dt** [comment]
Review status: 2 of 5 files reviewed at latest revision, all discussions resolved.

---

*[sql/group.go, line 53 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7DulQ4S48glBi98VSe:-K7DxHPlxow1bH8Yo_uo:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/group.go#L53)):*
Done.

---

*[sql/scan.go, line 524 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7DtvAYuy9ViaXNcqaV:-K7DxJBpU645ODa2KADl:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd9fe4ae0d2383fbeaa7e842bb30fdcbbf3449e/sql/scan.go#L524)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Comment by tamird** [comment]
Reviewed 3 of 3 files at r1, 2 of 2 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[sql/group.go, line 238 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJbhn4s0JBmv1q-bp:-K7EJbhn4s0JBmv1q-bq:-733388523) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/group.go#L238)):*
did this actually matter?

---

*[sql/group.go, line 53 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EKMZg7Yjgf7wFvY8C:-K7EKMZhKUoNzTskvxUt:-561417835) ([raw file](https://github.com/cockroachdb/cockroach/blob/e32fb6a7e3a756dec8abc8d5cf4df2a0187f5f6e/sql/group.go#L53)):*
confused again, isn't `col` 1-indexed? shouldn't the check be that `col > 0`?

---

*[sql/group.go, line 574 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EK9dj4_bjjqO3eq3b:-K7EK9dj4_bjjqO3eq3c:1894916111) ([raw file](https://github.com/cockroachdb/cockroach/blob/e32fb6a7e3a756dec8abc8d5cf4df2a0187f5f6e/sql/group.go#L574)):*
s/ie/i.e./

---

*[sql/group.go, line 576 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EKF3pCyeRukogZP7w:-K7EKF3pCyeRukogZP7x:1089441287) ([raw file](https://github.com/cockroachdb/cockroach/blob/e32fb6a7e3a756dec8abc8d5cf4df2a0187f5f6e/sql/group.go#L576)):*
stick to Eval rather than mixing case.

---

*[sql/scan.go, line 394 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EIW99hQzJIsLDumpj:-K7EIW99hQzJIsLDumpk:-766706638) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd9fe4ae0d2383fbeaa7e842bb30fdcbbf3449e/sql/scan.go#L394)):*
is this actually needed? originalRender is not used at all, and originalCols is used but I'm not convinced it's necessary. Can you make this comment more specific (i.e. what happens in group to make it needed?)

---

*[sql/scan.go, line 530 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EI4GSRrGPppKA2Dbr:-K7EI4GSRrGPppKA2Dbs:1405115165) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd9fe4ae0d2383fbeaa7e842bb30fdcbbf3449e/sql/scan.go#L530)):*
this looks like a change in behaviour; is it? needs a test if so

---

*[sql/scan.go, line 533 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJB_6l6uZ1V1Yzfaa:-K7EJB_6l6uZ1V1Yzfab:1155359453) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/scan.go#L533)):*
why this special case? perhaps clear in the next commit.

---

*[sql/scan.go, line 534 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EIpfBPhdJHuJwsVXv:-K7EIpfBPhdJHuJwsVXw:801375641) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/scan.go#L534)):*
i don't think this is the way to check for constants; what if someone writes `select 1 order by true && false`?

---

*[sql/sort.go, line 93 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJQ7KFiFP54O199DJ:-K7EJQ7KFiFP54O199DK:-1059738127) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/sort.go#L93)):*
why the +1? isn't this return value already 1-indexed?

---

*[sql/testdata/aggregate, line 32 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EIdlFQqr_DxeBgmo9:-K7EIdlFQqr_DxeBgmoA:-689312852) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/testdata/aggregate#L32)):*
test the error cases?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Comment by dt** [comment]
Review status: 3 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/group.go, line 238 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJbhn4s0JBmv1q-bp:-K7HXB-I2EMCLH1khlUz:1588918014) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/group.go#L238)):*
yes, moving it to after the bucket filling is what changes the `Eval` behavior in `aggregateFunc`

---

*[sql/group.go, line 53 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EKMZg7Yjgf7wFvY8C:-K7HY6We8kktwLUvST62:-1297126015) ([raw file](https://github.com/cockroachdb/cockroach/blob/e32fb6a7e3a756dec8abc8d5cf4df2a0187f5f6e/sql/group.go#L53)):*
`colIndex` returns it 0-indexed so it can be used directly with `s.render`, and so that we're never passing around a 1-indexed offset which might be confusing.

---

*[sql/group.go, line 574 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EK9dj4_bjjqO3eq3b:-K7HY31QtEjXKT_jaVID:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e32fb6a7e3a756dec8abc8d5cf4df2a0187f5f6e/sql/group.go#L574)):*
Done.

---

*[sql/group.go, line 576 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EKF3pCyeRukogZP7w:-K7HY2hfcBxqnvXal9c4:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/e32fb6a7e3a756dec8abc8d5cf4df2a0187f5f6e/sql/group.go#L576)):*
Done.

---

*[sql/scan.go, line 394 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EIW99hQzJIsLDumpj:-K7HYl101DjQ2fQ0Tfdd:2036712539) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd9fe4ae0d2383fbeaa7e842bb30fdcbbf3449e/sql/scan.go#L394)):*
Clarified in the comment. `originalRender` was indeed unused (for now) so I killed it.

---

*[sql/scan.go, line 530 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EI4GSRrGPppKA2Dbr:-K7H_EzI6Lii6Bhkc0Nx:-429879493) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd9fe4ae0d2383fbeaa7e842bb30fdcbbf3449e/sql/scan.go#L530)):*
Hm, it shouldn't be -- I just wanted to be able to pass the raw `GROUP BY` expressions to this before going to the effort of resolving and normalizing them, but that means they'll be IntVals, not DInts, when I see them.

---

*[sql/scan.go, line 533 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJB_6l6uZ1V1Yzfaa:-K7HasJYIqhhJah5PVIx:279254451) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/scan.go#L533)):*
GROUP BY "hello

---

*[sql/scan.go, line 534 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EIpfBPhdJHuJwsVXv:-K7HaeodmgTObylIZlY5:1364737350) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/scan.go#L534)):*
interestingly, postgres allows that, because I think it casts bools to ints? 

Maybe I should go ahead and normalize first, which would do constant folding?

fwiw, postgres errors with `non-integer constant` on `select 1 group by 'hello';` but seems fine with `select 1 group by upper('hello')`, so it seems like they do a pretty simple check for constants.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Comment by tamird** [comment]
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/group.go, line 238 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJbhn4s0JBmv1q-bp:-K7HeU4BuXOm2uNfZbvM:-975976067) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/group.go#L238)):*
OK, that's the consequence of moving it, but what was the impetus?

---

*[sql/group.go, line 53 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EKMZg7Yjgf7wFvY8C:-K7HeqnOmtUGmepon68K:1599802307) ([raw file](https://github.com/cockroachdb/cockroach/blob/e32fb6a7e3a756dec8abc8d5cf4df2a0187f5f6e/sql/group.go#L53)):*
`colIndex` returns a 0-indexed integer? I don't think so, since it blindly returns whatever the user passed, which is 1-indexed. What am I missing?

---

*[sql/scan.go, line 394 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EIW99hQzJIsLDumpj:-K7HdUtNhTucHAYZX-bb:-96179061) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd9fe4ae0d2383fbeaa7e842bb30fdcbbf3449e/sql/scan.go#L394)):*
That comment is a bit convoluted; I would just say "`GROUP BY` may internally add render targets, which we do not consider in ..."

---

*[sql/scan.go, line 530 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EI4GSRrGPppKA2Dbr:-K7Hdph6nfvr15rE0D0u:1029543405) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd9fe4ae0d2383fbeaa7e842bb30fdcbbf3449e/sql/scan.go#L530)):*
Add a comment explaining why we may see one or the other here.

---

*[sql/scan.go, line 533 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJB_6l6uZ1V1Yzfaa:-K7He2lAaZHKxB4TvcJz:-273938101) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/scan.go#L533)):*
postgres:
```sql
test=# select 1 order by 'hello';
ERROR:  non-integer constant in ORDER BY
LINE 1: select 1 order by 'hello';
                          ^
test=# select 1 group by 'hello';
ERROR:  non-integer constant in GROUP BY
LINE 1: select 1 group by 'hello';
                          ^
test=#
```

we should emit the same error in both cases AFAICT

---

*[sql/scan.go, line 534 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EIpfBPhdJHuJwsVXv:-K7HeGC7Y7JhmlQYltV7:963239375) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/scan.go#L534)):*
that's very strange. I think we can and should do better by normalizing first.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

File: sql/group.go [hunk 1]
```diff
@@ -229,6 +236,8 @@ func (n *groupNode) computeAggregates() {
 		n.buckets[""] = struct{}{}
 	}
 
+	n.intialized = true
```
Comment (tschottdorf) [inline sql/group.go]: s/int/init/?

**Comment by dt** [comment]
Review status: 0 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[sql/scan.go, line 394 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EIW99hQzJIsLDumpj:-K7Hw_9jR7Ar4WGC1Xu7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd9fe4ae0d2383fbeaa7e842bb30fdcbbf3449e/sql/scan.go#L394)):*
Done.

---

*[sql/scan.go, line 530 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EI4GSRrGPppKA2Dbr:-K7Hwptq4NYPJK6EEKHt:483958621) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dd9fe4ae0d2383fbeaa7e842bb30fdcbbf3449e/sql/scan.go#L530)):*
Just doing normalization first now so I removed it.

---

*[sql/scan.go, line 533 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJB_6l6uZ1V1Yzfaa:-K7Hwu3IDN5S0sdLgI5R:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/scan.go#L533)):*
Done.

---

*[sql/scan.go, line 534 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EIpfBPhdJHuJwsVXv:-K7Hwv-TzZ_DzguqEHc0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/scan.go#L534)):*
Done.

---

*[sql/sort.go, line 93 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJQ7KFiFP54O199DJ:-K7Hd9ElsMd7kekeqX1T:-1761635292) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/sort.go#L93)):*
no, `s.colIndex` is 0-indexed, to match `s.render`.

---

*[sql/testdata/aggregate, line 32 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EIdlFQqr_DxeBgmo9:-K7Hx0YAMifuXkHj4Kr1:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/testdata/aggregate#L32)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Comment by tamird** [comment]
Reviewed 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/sort.go, line 93 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJQ7KFiFP54O199DJ:-K7I-nLXi_N0dDNRDeOs:-2077123418) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/sort.go#L93)):*
See comment in group.go; I don't see how `colIndex` is 0-indexed.

---

*[sql/testdata/aggregate, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7I-z6Rwhfdg5Ti4rMd:-K7I-z6Rwhfdg5Ti4rMe:-983649386) ([raw file](https://github.com/cockroachdb/cockroach/blob/525d758c8fe2b0fb8924b60fab7f24fa205f144f/sql/testdata/aggregate#L42)):*
you should restore this as an error case

---

*[sql/testdata/aggregate, line 59 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7I-va1vJK6ooXm9y0u:-K7I-va1vJK6ooXm9y0v:-113006217) ([raw file](https://github.com/cockroachdb/cockroach/blob/91391e884dadda504f5275962b7d1c9b0d6630c6/sql/testdata/aggregate#L59)):*
spurious newline

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Comment by dt** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/group.go, line 238 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJbhn4s0JBmv1q-bp:-K7I-9uWSuWrVKaF4fYE:-7662183) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/group.go#L238)):*
Usually we just append GROUP BY expressions, normalized but otherwise mostly unchanged, to `s.render`. During bucket filling, we evaluate and concatenate those to come up with the bucket key.

However, when we replace `GROUP BY 2` with GROUP BY s.render[1]`, since we're reusing the expression that `extractAggFuncs` will modify (swapping its qvalues  to aggregatefuncs). When we try to eval _that_ to get the bucket key it will not yet have its value (since values are added *per-bucket*), so it fails. This change makes the aggregateFunc pass Eval through to the underlying, pre-wrapped, expression until bucket filling is complete, so that bucket-key generation will work.

The alternative would be to have the replacement of `GROUP BY 2` make a deep copy of `s.render[1]`  (which would not get changed to use `aggregateFuncs`) and could be esaily renderable for the purposes of picking buckets, but AFAIK we don't have a deepCopy of arbitrary Expr trees and it seemed non-trivial to add.

---

*[sql/group.go, line 53 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EKMZg7Yjgf7wFvY8C:-K7I-5lAVW-JiQvIHmal:31412062) ([raw file](https://github.com/cockroachdb/cockroach/blob/e32fb6a7e3a756dec8abc8d5cf4df2a0187f5f6e/sql/group.go#L53)):*
`return index - 1, nil`

---

*[sql/group.go, line 239 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7DrBXSRMxUGo5-fBQy-r4-239:-K7I-9TKt9hQ97r9zorA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/525d758c8fe2b0fb8924b60fab7f24fa205f144f/sql/group.go#L239)):*
Done.

---

*[sql/sort.go, line 93 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJQ7KFiFP54O199DJ:-K7I2PMi-aAoXtZJIzyg:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/sort.go#L93)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Comment by tamird** [comment]
LGTM

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/group.go, line 238 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJbhn4s0JBmv1q-bp:-K7IF9nnq4SuqmtYrklg:-2107306620) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/group.go#L238)):*
Cool, that's subtle. Can you add this detailed explanation somewhere in the code?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-:-K7IFKHbUPAnTHTSs5wU:676834233)*

**Comment by dt** [comment]
Review status: 2 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/group.go, line 238 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7EJbhn4s0JBmv1q-bp:-K7IREjc7EW3LPy1mRLV:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/5e6fd661ea1525f192ba8789c52fe9b158fef784/sql/group.go#L238)):*
Done.

---

*[sql/testdata/aggregate, line 42 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7I-z6Rwhfdg5Ti4rMd:-K7IRSF5Qm73jIzaXh1T:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/525d758c8fe2b0fb8924b60fab7f24fa205f144f/sql/testdata/aggregate#L42)):*
Done.

---

*[sql/testdata/aggregate, line 59 \[r5\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7I-va1vJK6ooXm9y0u:-K7IRTa5v8E270p2nNrj:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/91391e884dadda504f5275962b7d1c9b0d6630c6/sql/testdata/aggregate#L59)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Comment by tamird** [comment]
Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/scan.go, line 524 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7IRu_9b4CfMAxMWD0M:-K7IRu_AeaRXLviycLlb:155189179) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bb99e102816421aec71d9ea4554db2c9ff0f4ac/sql/scan.go#L524)):*
remove this space

---

*[sql/scan.go, line 525 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7IRz5IyZpb026n09i6:-K7IRz5IyZpb026n09i7:-1021697086) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bb99e102816421aec71d9ea4554db2c9ff0f4ac/sql/scan.go#L525)):*
nit: this is still ok as a type switch - the current impl will need to grab the type of `expr` multiple times.

---

*[sql/scan.go, line 527 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7ISKAXvqtw0563rYGT:-K7ISKAXvqtw0563rYGU:-501075525) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bb99e102816421aec71d9ea4554db2c9ff0f4ac/sql/scan.go#L527)):*
nit: save a call to `len` here: `if numCols := len(n.originalCols); index < 1 || index > numCols { .... fmt.Errorf(..., numCols) }`

---

*[sql/scan.go, line 538 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7ISCykcl2FMTh5M7n9:-K7ISCykcl2FMTh5M7nA:1610652954) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bb99e102816421aec71d9ea4554db2c9ff0f4ac/sql/scan.go#L538)):*
throw a comment on this or the function explaining what this return signals

---

*[sql/testdata/aggregate, line 50 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7ISk2xwX1YwCV9qJ_o:-K7ISk2xwX1YwCV9qJ_p:-1168052600) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bb99e102816421aec71d9ea4554db2c9ff0f4ac/sql/testdata/aggregate#L50)):*
nit: "non-integer constant column index"

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Comment by dt** [comment]
Review status: 2 of 5 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/scan.go, line 524 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7IRu_9b4CfMAxMWD0M:-K7IWqjT3yZ4lgnfhrmo:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bb99e102816421aec71d9ea4554db2c9ff0f4ac/sql/scan.go#L524)):*
Done.

---

*[sql/scan.go, line 525 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7IRz5IyZpb026n09i6:-K7IXKvGBobfQEz14XnT:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bb99e102816421aec71d9ea4554db2c9ff0f4ac/sql/scan.go#L525)):*
Done.

---

*[sql/scan.go, line 527 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7ISKAXvqtw0563rYGT:-K7IXudDaHV-XPXrXsi0:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bb99e102816421aec71d9ea4554db2c9ff0f4ac/sql/scan.go#L527)):*
Done.

---

*[sql/scan.go, line 538 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7ISCykcl2FMTh5M7n9:-K7IXw9yj3v6I6D0VcBt:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bb99e102816421aec71d9ea4554db2c9ff0f4ac/sql/scan.go#L538)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Comment by dt** [comment]
Review status: 2 of 5 files reviewed at latest revision, all discussions resolved.

---

*[sql/testdata/aggregate, line 50 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597#-K7ISk2xwX1YwCV9qJ_o:-K7IY3ZlY6oOTCNNQzIm:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bb99e102816421aec71d9ea4554db2c9ff0f4ac/sql/testdata/aggregate#L50)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Comment by tamird** [comment]
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3597)*

**Pull request merged**